File: DialPad/NewPassword/NewPassword/ViewController.m
//
//  ViewController.m
//  NewPassword
//
//  Created by Evo-Mac on 12/04/20.
//  Copyright © 2020 Evo-Mac. All rights reserved.
//

#import "ViewController.h"
#import <UIKit/UIKit.h>
@interface ViewController ()

@end

@implementation ViewController


- (void)viewDidLoad {
    [super viewDidLoad];
    
    NSString *newPassword =[self newPassword:@"ab" b:@"zsd"];
    NSLog(@"NewPassword is %@",newPassword);
    
    
//    NSArray *items =[[NSArray alloc]initWithObjects:@"1",@"1",@"1",@"2",@"2",@"3",nil];
//    NSNumber *number = [NSNumber numberWithInteger:2];
//    [self deleteProducts:items m:number];
    
    
    // Do any additional setup after loading the view.
}


- (NSString *) newPassword:(NSString *)a b:(NSString *)b {
    // Write your code here
    
    NSUInteger maxStringLength;
    NSString *resultStr = nil;
    resultStr=[NSString stringWithFormat:@""];
    char fc ,sc;
    NSString *fcStr,*scStr;
    
    
    if(a.length > b.length)
    {
        maxStringLength = a.length;
    }
    else if(b.length > a.length)
    {
        maxStringLength = b.length;
    }
    else {
        maxStringLength = a.length;
    }
    
    
    
    
    
    for (int i=0; i<maxStringLength; i++)
    {
        if(a.length > i)
        {
            fc = [a characterAtIndex:i];
            fcStr =[NSString stringWithFormat:@"%c",fc];
        }
        else
        {
            fcStr=@"";
        }
   
        
        if(b.length > i)
        {
            sc = [b characterAtIndex:i];
            scStr =[NSString stringWithFormat:@"%c",sc];
        }
        else
        {
            scStr=@"";
        }
 

        resultStr=[resultStr stringByAppendingString:fcStr];
        resultStr=[resultStr stringByAppendingString:scStr];
        
    }
    
    return resultStr;
}



/*
- (NSNumber *) deleteProducts:(NSArray *)ids m:(NSNumber *)m {
    // Write your code here
    
    NSArray *uniqueObjects =[[NSSet setWithArray:ids]allObjects];
    NSLog(@"uniqueObjects from array %@",uniqueObjects);
    NSLog(@"uniqueObjects Count is %lu",(unsigned long)uniqueObjects.count);
    
    
    
    for (int i = 0; i<uniqueObjects.count; i++) {
        
        NSString *unqieObject = [NSString stringWithFormat:@"%@",[uniqueObjects objectAtIndex:i]];
        
        
        NSLog(@"unqieObject is %@",unqieObject);
        int uniqueIDItemCountInTheBag = 0;

        for (int j = 0; j<ids.count; j++)
        {
            NSString * itemFromBag =[NSString stringWithFormat:@"%@",[ids objectAtIndex:j]];
            
            if([unqieObject isEqualToString:itemFromBag])
            {
                uniqueIDItemCountInTheBag++;
                
                //NSLog(@"Items count with ID %@ in bag is %d",unqieObject,uniqueIDItemCountInTheBag);
               // NSLog(@"Item with this uniqueID is %@ and ID is %@",itemFromBag,unqieObject);
            }
            else
            {
                NSLog(@"Items count with ID %@ in bag is %d",unqieObject,uniqueIDItemCountInTheBag);

            }
        }
    }
    
    
    NSNumber *result =[NSNumber numberWithInteger:3];
    
    return result;
    
}
*/


@end

File: General/MergeSort/MergeSort/ViewController.swift
//
//  ViewController.swift
//  MergeSort
//
//  Created by Evo-Mac on 22/04/20.
//  Copyright © 2020 Evo-Mac. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        let resultArray = self.mergeSort(array: [1,4,18,7,9,2,5,0])
        print(resultArray)
    }
    

    
    
   public func mergeSort<Element>(array : [Element]) -> [Element] where Element : Comparable
    {
        
        guard array.count > 1 else {
            
            return array
        }
        
        let middle = array.count/2
        
        let leftArray = mergeSort(array: Array(array[..<middle]))
        let rightArray = mergeSort(array: Array(array[middle...]))
        
//        print("leftarray\(leftArray)")
//        print("rightArray\(rightArray)")

        return merge(leftArray: leftArray, rightArray: rightArray);
    }
    
    private func merge<Element>(leftArray : [Element], rightArray:[Element])-> [Element] where Element : Comparable
    {
        
        var leftIndex = 0
        var rightIndex = 0
        
        var result : [Element] = []
        
        while leftIndex < leftArray.count && rightIndex < rightArray.count
        {
            
            let leftElement  = leftArray[leftIndex]
            let rightElement = rightArray[rightIndex]
            
            if leftElement < rightElement
            {
               result.append(leftElement)
                leftIndex += 1
            }
            else if leftElement > rightElement
            {
                result.append(rightElement)
                rightIndex += 1
            }
            else
            {
                result.append(leftElement)
                leftIndex += 1
                result.append(rightElement)
                rightIndex += 1
            }
        }
        
        if leftIndex < leftArray.count
        {
            result.append(contentsOf: leftArray[leftIndex...])
            
        }
        
        if(rightIndex < rightArray.count)
        {
            result.append(contentsOf:rightArray[rightIndex...])
        }
        
        
        return result
        
    }
    
    
    
    

}

